### DISCUSSION TITLE
Avalonia and Webview2 - got it partially working ...

### DISCUSSION THREAD

**Discussion by kerriganb** [AvaloniaUI/Avalonia#8001] [General]
Hey all, ... I have Webview2 partially working in Avalonia ...Weirdly, if there is no XAML other than webview2 control - it renders perfectly, however if you add some xaml above or below it, it fails to render

this works perfectly
```
<Window xmlns="https://github.com/avaloniaui"
... >
	<Grid    Background="BlueViolet">
		<local:EmbedSample />
	</Grid>
</Window>
```
but weirdly ...this fails
```
<Window xmlns="https://github.com/avaloniaui" 
...>
	<Grid RowDefinitions="Auto * Auto">
		<Grid HorizontalAlignment="Stretch" VerticalAlignment="Top" Margin="6,6,6,6"  Grid.Row="0" >
			<Label Content="WebView2"  FontSize="14" FontWeight="Bold" VerticalAlignment="Center" HorizontalAlignment="Center"/>
		</Grid>
		<Grid  Grid.Row="1" Background="BlueViolet">
			<local:EmbedSample />
		</Grid>
		<Grid Grid.Row="2" >
			<Label Content="by Foobar"  FontSize="12" FontWeight="Bold" VerticalAlignment="Center" HorizontalAlignment="Center"/>
		</Grid>
	</Grid>
</Window>
```
.. it's almost like xaml content is drawing over the webview2 content - but looking at devtools - it seems like it's not ...

i'm attaching a sample project - ...

please - any help appreciated 

thank you,
Kerrigan.
[AvaloniaApplicationEmbed.zip](https://github.com/AvaloniaUI/Avalonia/files/8500369/AvaloniaApplicationEmbed.zip)

**Comment by kerriganb** [comment]
oh wow, typically happens ..you ask a question - they you realize how stupid you are ...

works now - was passing hParent (avalonia) - but i was never assigning or showing it!!

look at attached zip - goto TestDLL ... and pass in the parent Avalonia window ... they show it ...

		HWND hWnd = CreateWindow(
			szWindowClass,
			szTitle,
			WS_POPUP,
			CW_USEDEFAULT, CW_USEDEFAULT,
			1200, 900,
			hParent,   <--- was null before - doh!
			NULL,
			hInstance,
			NULL
		);

now show it

		ShowWindow(hParent,
			SW_SHOW);
		UpdateWindow(hParent);

works! 

though next step is to figure out the lag when you resize ...anyone has an idea on that ? it's the typical childwindow issue i think ...

**Reply by maxkatz6** [reply]
> though next step is to figure out the lag when you resize ...anyone has an idea on that ? it's the typical childwindow issue i think ...

I think it's typical childwindow AND webview issue. Even big browsers quite laggy on resizing.

**Comment by maxkatz6** [comment]
Just for example, you also can do it with existing .NET wrapper - Microsoft.Web.WebView2. It will make embed control quite simple to use.
But obviously your solution with unmanaged code is way cheaper - it doesn't require including heavy WebView2 library from nuget + winforms dependency.
Perfect solution would be to write own COM interop with WebView (using MicroCOM or specially configured CsWinRT as a glue), but it will be very time consuming. Though unmanaged library as a glue will go too.

[WebView2Sample.zip](https://github.com/AvaloniaUI/Avalonia/files/8500673/WebView2Sample.zip)

**Reply by kerriganb** [reply]
ah nice, didn't know about that ... i'll give it a shot...

btw: kinda unrelated ... but have you noticed (likely your solution has same problem?) ...that because of the interop helper ...it creates a child window ...and because of this ... there seems to be a lag ...when the main avalonia window is moved...

weirdly - i would think that - size would bypass messagequeue and send message direct - i.e. so you do not have delay when main window is moved (if you move fast you see it draw!) ... i was thinking i would capture the size event on avalonia window - then call moveWindow in win32

weirdly - it seems that handleResized is not sealed in avalonia ! ... any way we can get access to resize event?

thanks
Kerrigan.

**Reply by kerriganb** [reply]
sorry typo ... now sealed ... though docs seem to suggest it can override ...

**Reply by timunie** [reply]
Maybe related to the airspace issue which is also known from WPF for example

**Reply by longhibianca** [reply]
Hi @kerriganb . To use webview2 with avalonia, did you follow similar steps to webview2 for wpf (https://learn.microsoft.com/pt-br/microsoft-edge/webview2/get-started/wpf)? I'm trying to test webview2 with avalonia, but I didn't found any documentation to help me

**Comment by kerriganb** [comment]
Hi , there are 2 ways to consume webview2 . The easier way via c# wrapper , which unfortunately makes it a big install but MUCH easier to consume  or via c++ interop.

I am unfortunately out of country right now , but will give you examples of both when I get back in a couple of days.

Kerrigan

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by longhibianca** [reply]
Hi, thanks! I'll wait for you to get back to me.

**Comment by kerriganb** [comment]
Hi Bianca , I have some some code , but did you look at the attached sample from maxkatz6 higher in thread? That used to work. I also have my own c++ version of you like , though I would have to pair it down a bit for you . It has a lot of other code in it.

I just tried this other example and it still worked . Give it a shot and if you still want my c++ version I can send to you .

Try this link :

https://github.com/BeyondDimension/Avalonia.WebView2


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by longhibianca** [reply]
Hi kerriganb! No, I haven't seen the sample, but I'll look. No problem, I think maxkatz6's example should help me.
I've seen this package from beyond dimension, but there's a comment saying that it doesn't work with avalonia 11. When you tested it, was it with version 11 of avalonia or an earlier version?
